### ISSUE TITLE
Gracefullly terminate upon Ctrl-C

### ISSUE THREAD

**Issue by bangerth** [geodynamics/aspect#1357]
Sometimes it would be nice if one could terminate a job that doesn't want to converge and still get some kind of information out of the program where it currently is. An example would be to flush the convergence history of the solver to disk, which is currently only done upon postprocessing, but if a computation doesn't get there, we will never get this on disk.

Unix doesn't offer all that much that would allow us to see where a program is, but we could install a signal handler that at least allows us to run an emergency signal to which different parts of the program could attach.

Here's a working example:
```

struct X
{
  void doit ()
    {
      for (unsigned int i=0; i<100; ++i)
        {
          std::cout << i << std::endl;
          sleep (1);
        }
    }

  void error_handler (int sig)
    {
      std::cout << "Signal " << sig << " received." << std::endl;
    }
};


X *main_object;


void sighandler (int sig)
{
  main_object->error_handler(sig);
  exit (1);
}




int main ()
{
  struct sigaction sa;

  X x;
  main_object = &x;

  sa.sa_handler = sighandler;
  sigemptyset(&sa.sa_mask);
  sa.sa_flags = SA_RESTART; /* Restart functions if
                                 interrupted by handler */
  int err = sigaction(SIGINT, &sa, NULL);
  
  x.doit ();
}
```

What do people think?

**Comment by jaustermann** [comment]
I think this would be helpful! For some of the large viscosity variation runs the convergence is slow and it has definitely happened that runs timeout without output. If there would be an option to output the convergence history of the solver I would use it. 

~ Jacky

**Comment by jaustermann** [comment]
Hi group,

Has anyone done / tried the Zhong et al. (2008) G3 benchmarks for surface and CMB topography? If this hasn’t been done and people think it’s worthwhile I’d be happy to give it a shot. This might also be a possibility to test the geoid post processors that people worked on during the last hackathon. Ian and Shangxin, if you’re willing to share I could test your geoid post processors (just let me know if the ones that are on your github repositories are up to date or update them if they aren’t).

Cheers,
Jacky

**Comment by jaustermann** [comment]
Hi Jacky and Ian,

Last Month I found that my geoid code may still have bugs because I failed
to reproduce the geoid with the same rheology from CitcomS. I already
confirmed that my topography part has no problem (same as the output grid
topography when computed back from spherical harmonics) but the density
integral part still needs to be tested and I'm working on this now, also
including reproducing the geoid kernel in Zhong et al. (2008) paper. I'd
like to also try one calculation with the newly merged true boussinesq
approximation after I update my ASPECT code. I haven't benchmarking the
topography calculation with the paper. That's of course also a worthwhile
work.

There is no essential change in my geoid code on github. I only add some
extra output such as splitting the density and topography contribution in
my own code. If you need this, I can further update it.

Best,
Shangxin

**Comment by bangerth** [comment]
Hi Jacky & Shangxin -- would you mind having this conversation on the mailing list again (just repost your mails) instead of attaching it to this github issue? ;-))

**Comment by bangerth** [comment]
(I do think having geoid codes in ASPECT would be great!)

**Comment by jaustermann** [comment]
Ups, sorry!!

**Comment by tjhei** [comment]
> What do people think?

I am concerned how signal handling will interact with c++ (constructors/destructors/leaks etc.) and MPI.

**Comment by bangerth** [comment]
Yes, me too. But there is *already* no graceful recovery if someone hits Ctrl-C. I'm not proposing something that would yield a graceful recovery either. All I want is to be able to call a signal that allows parts of the code to do something, knowing full well that there is no guarantee at all that the data is correct, that we are at a state where the data is consistent, or that what the function does will actually be executed and written to disk. I just want to have some kind of probability that something useful will actually happen :-)
